text_classification.py

```diff
@@ -0,0 +1,487 @@
+
+             - AG_NEWS
+             - SogouNews
+             - DBpedia
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+
+    """
+
+    def __init__(self, dataset_name, root='.data',
+                 text_field=None, label_field=None, ngrams=1):
+        """Create a text classification dataset instance.
+
+        Arguments:
+            dataset_name: The name of dataset, include "ag_news", "sogou_news",
+                "dbpedia", "yelp_review_polarity", "yelp_review_full",
+                "yahoo_answers", "amazon_review_full", "amazon_review_polarity".
+            root: Directory where the dataset are saved. Default: ".data"
+            text_field: The field that will be used for the sentence. If not given,
+                'spacy' token will be used.
+            label_field: The field that will be used for the label. If not given,
+                'float' token will be used.
+            ngrams: a contiguous sequence of n items from s string text.
+                Default: 1
+
+        Examples:
+            >>> txt_cls = TextClassificationDataset("ag_news", ngrams=2)
+
+        """
+        self.dataset_name = dataset_name
+        if dataset_name in supported_text_classification_datasets.keys():
+            self.url = supported_text_classification_datasets[dataset_name]
+        else:
+            raise ValueError("dataset_name %s is not supported." % dataset_name)
+        self.root = root
+        fields = []
+        fields.append(('text', text_field if text_field is not None
+                       else data.Field(tokenize=data.get_tokenizer('spacy'),
+                                       init_token='<SOS>',
+                                       eos_token='<EOS>')))
+        fields.append(('label', label_field if label_field is not None
+                       else data.LabelField(dtype=torch.float)))
+        self.fields = dict(fields)
+
+        # Unpack field tuples
+        for n, f in list(self.fields.items()):
+            if isinstance(n, tuple):
+                self.fields.update(zip(n, f))
+                del self.fields[n]
+
+        self.raw_folder = os.path.join(self.root, self.__class__.__name__, 'raw')
+        self.processed_folder = os.path.join(self.root,
+                                             self.__class__.__name__,
+                                             'processed')
+
+        filepath = os.path.join(self.processed_folder, self.dataset_name + '.train')
+        self.train_examples = self.load_text_classification_data(filepath, ngrams)
+
+        filepath = os.path.join(self.processed_folder, self.dataset_name + '.test')
+        self.test_examples = self.load_text_classification_data(filepath, ngrams)
+        self.examples = self.train_examples + self.test_examples
+
+        self.fields['text'].vocab = build_vocab(self, self.fields['text'], 'text')
```

Comment:
I think we can call into this using self.examples. Then build_vocab can be more generic . This however also requires us to get rid of "fields".

Reply:
I will try to improve the build_vocab API. For now, I think two separate components, data (e.g. examples) and token processor (e.g. field) make sense.